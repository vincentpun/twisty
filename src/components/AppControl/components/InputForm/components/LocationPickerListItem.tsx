import * as React from 'react';
import * as classNames from 'classnames';

interface LocationListItemProps {
  title: string;
  active?: boolean;
  onItemClick?: () => void;
  rightAccessoryRender?: () => JSX.Element;
}

const LocationPickerListItem = ({ title, active, onItemClick, rightAccessoryRender }: LocationListItemProps) => (
  <li
    className={classNames(
      'location-picker__list-item',
      {
        'location-picker__list-item--selected': active,
        'location-picker__list-item--with-right-accessory': !!rightAccessoryRender,
      },
    )}
  >
    <div
      onClick={!!onItemClick ? () => onItemClick() : null}
      className="location-picker__list-item-text"
    >
      {title}
    </div>
    {
      !!rightAccessoryRender ?
      <span className="location-picker__list-item-right-accessory">
        <span className="location-picker__list-item-icon">
          {rightAccessoryRender()}
        </span>
      </span> :
      null
    }
  </li>
);

export default LocationPickerListItem;
