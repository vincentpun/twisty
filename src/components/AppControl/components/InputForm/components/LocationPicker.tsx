import * as React from 'react';
import * as classNames from 'classnames';

import LocationPickerListItem from './LocationPickerListItem';

interface LocationPickerProps {
  sectionTitle: string;
  titleAccessoryRender?: () => JSX.Element;
  items?: LocationPickerListItem[];
  onItemClick?: (item: LocationPickerListItem) => void;
}

export interface LocationPickerListItem {
  title: string;
  key: number | string;
  active?: boolean;
  rightAccessoryRender?: (item: LocationPickerListItem) => JSX.Element;
}

const LocationPicker = ({
  sectionTitle,
  titleAccessoryRender,
  items,
  onItemClick,
}: LocationPickerProps) => (
  <div className="location-picker">
    <div className="location-picker__label">
      <span className="location-picker__label-text">
        {sectionTitle}
      </span>
      {
        !!titleAccessoryRender ?
        <span className="location-picker__label-accessory">
          {titleAccessoryRender()}
        </span> :
        null
      }
    </div>
    {
      !!items && items.length > 0 ?
      <ul className="location-picker__list">
        {
          items.map((item) => (
            <LocationPickerListItem
              title={item.title}
              active={item.active}
              key={item.key}
              onItemClick={!!onItemClick ? () => onItemClick(item) : null}
              rightAccessoryRender={!!item.rightAccessoryRender ? () => item.rightAccessoryRender(item) : null}
            />
          ))
        }
      </ul> :
      null
    }
  </div>
);

export default LocationPicker;
