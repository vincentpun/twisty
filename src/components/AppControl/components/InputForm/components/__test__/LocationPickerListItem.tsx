import * as React from 'react';
import { shallow } from 'enzyme';

import LocationPickerListItem from '../LocationPickerListItem';

describe('LocationPickerListItem', () => {
  it('matches the snapshot', () => {
    const wrapper = shallow(<LocationPickerListItem
      title="test"
      active={false}
    />);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onItemClick when text is clicked', () => {
    const spy = jest.fn();
    const wrapper = shallow(<LocationPickerListItem title="" active={false} onItemClick={spy} />);
    wrapper.find('.location-picker__list-item-text').simulate('click');

    expect(spy).toHaveBeenCalled();
  });

  it('calls onItemClick when text is clicked', () => {
    const spy = jest.fn();
    const wrapper = shallow(<LocationPickerListItem title="" active={false} onItemClick={spy} />);
    wrapper.find('.location-picker__list-item-text').simulate('click');

    expect(spy).toHaveBeenCalled();
  });

  it('calls and renders accessory renderer', () => {
    const spy = jest.fn();
    spy.mockReturnValue(<div className="item-accessory">Test</div>);

    const wrapper = shallow(<LocationPickerListItem title="" rightAccessoryRender={spy} />);
    expect(wrapper.find('.item-accessory').length).toBe(1);
    expect(spy).toHaveBeenCalled();
  });
});
