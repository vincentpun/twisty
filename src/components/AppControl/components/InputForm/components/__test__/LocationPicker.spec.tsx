import * as React from 'react';
import { shallow } from 'enzyme';

import LocationPicker from '../LocationPicker';
import LocationPickerListItem from '../LocationPickerListItem';

describe('LocationPicker', () => {
  it('matches the snapshot', () => {
    const wrapper = shallow(
      <LocationPicker
        sectionTitle="Section title"
        items={[{ key: 'a', title: 'Test' }]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the list with items', () => {
    const items = [
      {
        title: 'Test',
        key: 'a',
      },
      {
        title: 'Test 2',
        key: 'b',
      }
    ];

    const wrapper = shallow(<LocationPicker sectionTitle="" items={items} />);
    expect(wrapper.find('ul').length).toBe(1);
    expect(wrapper.find(LocationPickerListItem).length).toBe(2);
  });

  it('does not render the list with no items', () => {
    const items: {title: string, key: string}[] = [];

    let wrapper = shallow(<LocationPicker sectionTitle="" items={items} />);
    expect(wrapper.find('ul').length).toBe(0);

    wrapper = shallow(<LocationPicker sectionTitle="" />);
    expect(wrapper.find('ul').length).toBe(0);
  });

  it('does not have accessroy element with no title accessory renderer', () => {
    const wrapper = shallow(<LocationPicker sectionTitle="" />);
    expect(wrapper.find('.location-picker__label-accessory').length).toBe(0);
  });

  it('calls and renders the title accessory renderer', () => {
    const spy = jest.fn();
    spy.mockReturnValue(<div className="accessory-render">Test</div>);

    const wrapper = shallow(<LocationPicker sectionTitle="" titleAccessoryRender={spy} />);

    expect(spy).toHaveBeenCalled();
    expect(wrapper.find('.accessory-render').length).toBe(1);
    expect(wrapper.find('.accessory-render').text()).toBe('Test');
  });
});
