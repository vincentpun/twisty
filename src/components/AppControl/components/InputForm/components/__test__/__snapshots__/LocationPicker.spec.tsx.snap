// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationPicker matches the snapshot 1`] = `
<div
  className="location-picker"
>
  <div
    className="location-picker__label"
  >
    <span
      className="location-picker__label-text"
    >
      Section title
    </span>
  </div>
  <ul
    className="location-picker__list"
  >
    <LocationPickerListItem
      key="a"
      onItemClick={null}
      rightAccessoryRender={null}
      title="Test"
    />
  </ul>
</div>
`;
